### ISSUE TITLE
Improvements to `alternating_update`

### ISSUE THREAD

**Issue by mtfishman** [mtfishman/ITensorNetworks.jl#60]
Here's an issue to track improvements to make to the general tensor network solver code, `alternating_update`, following up on #59.

- [ ] Remove the time argument `t` from `alternating_update`, since it only makes sense for solvers that implement time evolution like `tdvp`.
- [ ] Improve keyword argument code patterns based on: https://itensor.github.io/ITensors.jl/dev/DeveloperGuide.html#Keyword-Argument-Best-Practices
- [ ] Generalize `ProjTTN` types to a more general `ITensorNetworkCache` type (or specifying custom caching types which are relevant for different contraction backends), based on a contraction sequence tree.
- [ ] Replace `ProjTTNSum` with a more general lazy sum, based on the `Applied` type in `ITensors.LazyApply`.
- [ ] Generalize to accept `AbstractITensorNetwork`, and ensure the interface requirements are general enough to work for more general networks.
- [ ] Think more carefully about the interface and implementation for solving various linear equations, like `Ax ≈ λx`, `Ax ≈ b`, `x ≈ y`, etc. which shares caching structures, works with general tensor networks, contraction backends, etc.

**Comment by emstoudenmire** [comment]
- [ ] Switch default tdvp solver to `exponentiate`

**Comment by mtfishman** [comment]
Thanks, I added it.

**Comment by emstoudenmire** [comment]
Proposed redesign of internals:
  1. Eliminate outermost for loop over substep (see the function update_step in update_step.jl)
  2. Pass a data structure which says how to do a single (full) sweep over the system. This defines what a sweep means, as in nsweeps=3 does three of these.
  3. This data structure can carry an optional "payload" (named tuple?). Using tdvp as an example, the payload would store the explicit, signed and weighted time steps, already preprocessed. In the case of a PEPS it could store the row and column number.
  4. There is an optional, second type of callback function (not a solver) which can post process the payload before feeding it into the solver. (May not be necessary to add this #4 callback feature if the payloads are sufficiently preprocessed.)
  

Basically Lander's data structures become the definition of what a sweep is and are generalized to not just store the forward/reverse sign, but any data relevant to that specific point in the sweep.
